google/cloud/firestore_bundle/bundle.py

```diff
@@ -0,0 +1,439 @@
+        self.named_queries: Dict[str, NamedQuery] = {}
+        self.latest_read_time: Timestamp = Timestamp(seconds=0, nanos=0)
+        self._deserialized_metadata: Optional[BundledDocumentMetadata] = None
+
+    def add(
+        self,
+        document_or_query_name: Union[DocumentSnapshot, str],
+        query: Optional[Query] = None,
+    ) -> "FirestoreBundle":
+        """Adds a document or entire query to the bundle.
+
+        This function is offered for convenience / API parity with other SDKs.
+        Most users will get better outcomes from using the more specific,
+        `add_document` or `add_named_query` methods.
+
+        Args:
+            document_or_query_name (Union[DocumentSnapshot, str]): If supplied
+                as a `DocumentSnapshot`, this parameter routes control flow to
+                `add_document`, and the second parameter must be None. However,
+                if supplied as a `str`, then this parameter routes control flow
+                to `add_named_query`, and the second parameter must be supplied.
+            query_snapshot (Optional[Query]): Query to fully load and save to
+                the bundle. This parameter is required if the first parameter is
+                a string, and is required to be `None` if the first parameter
+                is a `DocumentSnapshot`.
+
+        Example:
+
+            from google.cloud import firestore
+            from google.cloud.firestore.firestore_bundle import FirestoreBundle
+
+            db = firestore.Client()
+            collection_ref = db.collection(u'users')
+
+            bundle = FirestoreBundle('my-bundle')
+
+            # Add an entire query. This will load and save every matching
+            # document.
+            bundle.add('all the users', collection_ref._query())
+
+            # Add a single document.
+            bundle.add(collection_ref.documents('some_id').get())
+
+        Returns:
+            FirestoreBundle: self
+
+        """
+        if isinstance(document_or_query_name, DocumentSnapshot):
+            assert query is None
+            self.add_document(document_or_query_name)
+        elif isinstance(document_or_query_name, str):
+            assert query is not None
+            self.add_named_query(document_or_query_name, query)
+        else:
+            raise ValueError(
+                "Bundle.add accepts either a standalone DocumentSnapshot or "
+                "a name (string) and a Query."
+            )
+        return self
+
+    def add_document(
+        self, snapshot: DocumentSnapshot, query_name: Optional[str] = None,
+    ) -> "FirestoreBundle":
+        """Adds a document to the bundle.
+
+        Args:
+            snapshot (DocumentSnapshot): The fully-loaded Firestore document to
+                be preserved.
+            query_name (Optional[str]): If provided, also establishes a link
+                between the provided document and the referenced query.
```

Comment:
Sorry I am doing a rush review here so I can't verify how everything works, but IMHO queries are atomic units and we can't add documents to them. The client will not surface the document if it doesn't match.

Reply:
Yes. The mapping from document to matched query name should be deducted from `add_named_query`, we should not expect clients to manually provide these mappings, if that is the intention here.

Reply:
Oh I see. I was mimicking the function signatures from Java, but maybe I misunderstood their intent. I'll make this change.